dynamiqs/sparse.py

```diff
@@ -0,0 +1,124 @@
+        elif isinstance(other, ArrayLike):
+            return self.to_dense() - other
+        elif isinstance(other, SparseQArray):
+            _check_compatible_dims(self.dims, other.dims)
+            return self._sub_sparse(other)
+
+        return NotImplemented
+
+    def _sub_sparse(self, other: SparseQArray) -> SparseQArray:
+        out_offsets_diags = dict(zip(self.offsets, self.diags))
+        for other_offset, other_diag in zip(other.offsets, other.diags):
+            if other_offset in out_offsets_diags:
+                out_offsets_diags[other_offset] -= other_diag
+            else:
+                out_offsets_diags[other_offset] = -other_diag
+
+        out_offsets = tuple(sorted(out_offsets_diags.keys()))
+        out_diags = jnp.array([out_offsets_diags[offset] for offset in out_offsets])
+
+        return SparseQArray(out_diags, out_offsets, self.dims)
+
+    def __rsub__(self, other: Array) -> Array:
+        return -self + other
+
+    def __mul__(self, other: Array | SparseQArray) -> Array | SparseQArray:
+        if isinstance(other, ScalarLike):
+            if other == 0:
+                diags = jnp.empty(0, self.shape[-1])
+                offsets = ()
+                return SparseQArray(diags, offsets, self.dims)
+            return SparseQArray(other * self.diags, self.offsets, self.dims)
+        elif isinstance(other, Array):
+            return self._mul_dense(other)
+        elif isinstance(other, SparseQArray):
+            _check_compatible_dims(self.dims, other.dims)
+            return self._mul_sparse(other)
+
+        return NotImplemented
+
+    def _mul_dense(self, other: Array) -> SparseQArray:
+        N = other.shape[0]
+        out_diags = jnp.zeros_like(self.diags)
+        for i, (self_offset, self_diag) in enumerate(zip(self.offsets, self.diags)):
+            start = max(0, self_offset)
+            end = min(N, N + self_offset)
+            other_diag = jnp.diagonal(other, self_offset)
+            out_diags = out_diags.at[i, start:end].set(
+                other_diag * self_diag[start:end]
+            )
+
+        return SparseQArray(out_diags, self.offsets, self.dims)
+
+    def _mul_sparse(self, other: SparseQArray) -> SparseQArray:
+        out_diags, out_offsets = [], []
+        for self_offset, self_diag in zip(self.offsets, self.diags):
+            for other_offset, other_diag in zip(other.offsets, other.diags):
+                if self_offset != other_offset:
+                    continue
+                out_diags.append(self_diag * other_diag)
+                out_offsets.append(other_offset)
+
+        return SparseQArray(jnp.stack(out_diags), tuple(out_offsets), self.dims)
+
+    def __rmul__(self, other: ArrayLike) -> Array:
+        return self * other
+
+
+def _check_compatible_dims(dims1: tuple[int, ...], dims2: tuple[int, ...]):
+    if dims1 != dims2:
+        raise ValueError(
+            f'QArrays have incompatible dimensions. Got {dims1} and {dims2}.'
+        )
```

Comment:
```suggestion
def _check_compatible_dims(dims1: tuple[int, ...], dims2: tuple[int, ...]):
    if dims1 != dims2:
        raise ValueError(
            f'QArrays have incompatible dimensions. Expected `(..., n, k)` and `(..., k, m)`. Got {dims1} and {dims2}.'
        )
```

Reply:
These are not `dims` of a matrix product, but the `dims` for the underlying tensor product. So they should be equal, and not `(..., n, k)` and `(..., k, m)`.

But we can add "Expected `dims1 == dims2`, but got {dims1} and {dims2}."